api/features/versioning/signals.py

```diff
@@ -0,0 +1,57 @@
+import hashlib
+import time
+
+from django.db.models.signals import post_init, post_save, pre_save
+from django.dispatch import receiver
+from django.utils import timezone
+
+from environments.tasks import rebuild_environment_document
+from features.versioning.models import EnvironmentFeatureVersion
+
+
+@receiver(post_save, sender=EnvironmentFeatureVersion)
+def add_existing_feature_states(
+    instance: EnvironmentFeatureVersion, created: bool, **kwargs
+):
+    if not created:
+        return
+
+    previous_environment_feature_version = instance.get_previous_version()
+    if not previous_environment_feature_version:
+        return
+
+    for feature_state in previous_environment_feature_version.feature_states.all():
+        feature_state.clone(
+            env=instance.environment, environment_feature_version=instance
+        )
+
+
+@receiver(pre_save, sender=EnvironmentFeatureVersion)
+def generate_sha(instance: EnvironmentFeatureVersion, **kwargs):
+    if not instance.sha:
+        instance.sha = hashlib.sha256(
+            f"{instance.environment.id}{instance.feature.id}{time.time()}".encode(
+                "utf-8"
+            )
+        ).hexdigest()
+
+
+@receiver(pre_save, sender=EnvironmentFeatureVersion)
+def update_live_from(instance: EnvironmentFeatureVersion, **kwargs):
+    if instance.published and not instance.live_from:
+        instance.live_from = timezone.now()
+
+
+@receiver(post_init, sender=EnvironmentFeatureVersion)
+def cache_fields(instance: EnvironmentFeatureVersion, **kwargs):
+    instance._init_fields = {"published": instance.published}
```

Comment:
We're essentially implementing our own version of LifecycleModel here, to avoid circular import issues. Right now, we only care about `published` but we can extend it as needed later on.